### ISSUE TITLE
Use bootstrap 4 for phx generator markup/css

### ISSUE THREAD

**Issue by chrismccord** [phoenixframework/phoenix#2778]

**Comment by michalmuskala** [comment]
Since phoenix does not include all of bootstrap anyway in the default generators maybe it would make sense to have something other than that? Maybe even keep generating the same markup, but with some custom css? I think including parts of bootstrap is only confusing to people - there's a lot of questions about that.

**Comment by chrismccord** [comment]
The original goals were to have something that looks nice out-of-the-box that folks can roll with as they are getting started or prototyping. So we are not married to bootstrap, but given its ubiquity, having a popular solution can be more helpful over some custom css and markup or another more obscure setup. I would be open to alternatives, assuming:

- Someone wants to own this markup and css update, including making sure our forms style properly
- It works across all major browsers
- Is responsive out of the box

I agree including only parts of bootstrap can be confusing, but bootstrap itself is often used piecemeal so it's not completely unexpected. For giving workshops and teaching around Phoenix, having the sane grid system and styles that a lot of folks (and me :) already know about is extremely useful.

**Comment by webdeb** [comment]
Since https://github.com/phoenixframework/phoenix/pull/2779 is merged.
It would make sense to provide a proper configuration layout based on the webpack guides here: https://getbootstrap.com/docs/4.0/getting-started/webpack/ 

And try to prevent the confusion by referencing to the guides of webpack / bootstrap..

Hmm, just thought about the --no-webpack option.. and.. don't know how to deal with bootstrap in that case then..

**Comment by inoas** [comment]
Maybe this can be revisited once natuve css grud are widely adopted by browser vendors.

Then I suggest to take a look at https://github.com/mblode/marx

Combined with native css grids / flexbox productivity is great and dependencies are low (no class/div soup).

**Comment by anticide** [comment]
May I suggest Bulma (https://bulma.io/)? Is the CSS framework of my choice after using Bootstrap for a long time, some advantages:

- Easy to learn (very intuitive classes).
- Responsive.
- Uses flexbox.
- No JavaScript.

Other CSS frameworks are too minimal (Pure, Milligram...) or too bloated with JavaScript and jQuery (Foundation, Semantic UI...). Bulma works just fine for me.

I already have some progress in templates. I'm not a good designer to be honest so... any feedback would be much appreciated! Here's how it looks like:

Desktop: https://image.ibb.co/cFGyHn/Screenshot_2018_3_5_3.png
Mobile: https://image.ibb.co/eYO0V7/Screenshot_2018_3_5_1.png
Mobile (burger expanded): https://image.ibb.co/ikjUOS/Screenshot_2018_3_5_2.png

Notable changes:
- Added Font Awesome 5.
- Added a navbar. I think this is a must on every webpage and could be handy to have some scaffold.
- Added nav burger for mobile, requires a few lines of JavaScript to work properly.
- Added link to Stack Overflow.
- Updated doc links to hexdocs.pm.
- Updated Mailing list to Phoenix Forum.

If everything is fine, then I'll take CSS update on forms.

**Comment by thiagomajesk** [comment]
Even though I've been using mostly bootstrap 4 and tailwind css.
I vote for Bulma as a simpler alternative considering its small footprint and simpler style.

**Comment by josevalim** [comment]
@anticide Keep in mind that our goal here is to provide the minimum necessary so a just generated Phoenix application looks decent. So for example, adding new font dependencies is a no-no. Let's keep it simple and light because the odds are, whatever we do, users will just remove it anyway. So let's optimize for removal.

Another thing to keep in mind is that we also need to update the phx.gen.html generators if we change frameworks and I would like to avoid that the maximum we can. We already changed them for Phoenix v1.3, changing them again is just too many changes in a short period of time.

However, if bootstrap4 also requires changing the templates, then there is nowhere else to run.

And great catch on updating the links!

TLDR - think about incremental improvements rather than a full redesign.

**Comment by webdeb** [comment]
What if we completely get rid of bootstrap like @michalmuskala suggested. For the index.html we could just add some css `<style>` and for generators there should be actually an api, so that people can come up with their own templates, and use the classes they prefer instead of cleaning up each time after generation. 

As a developer, I do not care about the default fancy styling of some magically generated stuff, more interestingly for me is how I can change this default behaviour.. 

Just my 2 cents

**Comment by josevalim** [comment]
> As a developer, I do not care about the default fancy styling of some magically generated stuff, more interestingly for me is how I can change this default behaviour..

Which is why bootstrap is in a single file that you can just remove. You are right: being able to remove is the most important bit here.

The choice here is not supposed to matter. If someone wants to write a custom .css file that would look decent and responsive, we are fine with that too. We only use bootstrap because that's the easiest way to do this.

**Comment by inoas** [comment]
The benefit of using [MarxCSS](https://github.com/mblode/marx) ([Demo](https://codepen.io/mblode/details/JdYbJj)) would be that everyone can throw it away as there are no classes being used and you don't pollute your DOM. As such it works well with Elm/ReactJS especially if you simply add native CSS grids/flexbox

**Comment by josevalim** [comment]
I like the idea behind MarxCSS because we can get something nice looking without touching the markup.

I am not worried about the Phoenix index.html because that's definitely a one-off page and we can add at least *some* CSS in there.

Can someone build a small prototype and let us know how the phx.gen.html index and edit pages would look like with MarxCSS?

**Comment by pragmaticivan** [comment]
If there are any thoughts to incorporate bootstrap 4 using SCSS, it would be a simple transition, since you can use mixins to handle the grid part of it. Ex: https://getbootstrap.com/docs/4.0/layout/grid/#example-usage

It would reduce the efforts of updating the markup on `index.html` as long it has enough identifiers (classes/ids) to allow the simple usage of it.

**Comment by chrismccord** [comment]
We shipped scss initially but dropped it because it included like 20MB of test stubs node_modules. Has the situation improved there?

**Comment by webdeb** [comment]
> We shipped scss initially but dropped it because it included like 20MB of test stubs node_modules.

What? ...compiling `node_sass` is the most enjoyable process :D

**Comment by webdeb** [comment]
MarxCSS was initially designed as a reset. Thats what you need anyway, so 👍

**Comment by anticide** [comment]
@josevalim makes sense, thank you.

Regardless of the CSS framework, I'd suggest moving the header tag out of the layout (e.g. to the generated page). This is something that you'll always remove:

https://github.com/phoenixframework/phoenix/blob/master/installer/templates/phx_web/templates/layout/app.html.eex#L16-L23

This way the layout looks cleaner IMO. Also I noticed that the default meta tag is missing `shrink-to-fit=no` property, the standard meta tag is:

```html
<meta name="viewport" content="width=device-width, initial-scale=1.0, shrink-to-fit=no">
```

More about this: https://bitsofco.de/ios-safari-and-shrink-to-fit/

**Comment by josevalim** [comment]
> Regardless of the CSS framework, I'd suggest moving the header tag out of the layout (e.g. to the generated page). This is something that you'll always remove:

Then let people remove it. Again, the focus is on generating enough for what Phoenix needs, and Phoenix does not need headers per page.

**Comment by rockwood** [comment]
I'd like to second the idea of switching to something that doesn't require specific css classes. I use Bootstrap for just about everything, but I can understand how a newcomer might be confused by all the generated bootstrap classes. I also think it's important to have the generated templates look professional out of the box, so I think we need something that provides good looking styles by default.

I built out examples of class-less generators using these three frameworks:

  * [Marx](https://github.com/mblode/marx)
  * [Tacit](https://github.com/yegor256/tacit)
  * [BareCSS](https://github.com/longsien/BareCSS)

I think Marx is the best choice since it's the least intrusive. Tacit is nice, but it wanted to change the body color and add rounded corners in places that looked a little funny. I really like BareCSS's approach, but I think it's form styles are a little too complicated.

Here are some screenshots of Marx: 

index.html
![marx-3](https://user-images.githubusercontent.com/659052/37011049-a65958b6-2131-11e8-8530-cf527c0348e3.png)

edit.html:
![marx-1](https://user-images.githubusercontent.com/659052/37010571-88a705e0-212f-11e8-8197-59ec299ec083.png)

show.html
![marx-2](https://user-images.githubusercontent.com/659052/37010552-72bd76d8-212f-11e8-9bde-7afecfbccc48.png)

This removes all classes from the generated form markup, and adds a couple custom classes for flashes, help text and the home page. I'm going to prefix these with `phx-` so it's obvious that they're coming from `phoenix.css`.

Let me know if you think this will work, and I'll put together a PR.

**Comment by raarts** [comment]
As a newcomer I can confirm often being confused by irrelevant additions. My preference would even be to do no styling at all, because that's irrelevant to what I'm trying to learn (the templating process).

**Comment by yordis** [comment]
@chrismccord dont you think is better to add that as an extra deps? And then you would configure your default generators?

I don't use bootstrap since Bootstrap since 2016. Also from maintainability perspective, Bootstrap likes to break every major release their code so that will means more open issues and `maybe` causing Phoenix release delay.

**Comment by josevalim** [comment]
We went with [miligram](https://milligram.io/) at the end of the day:

![screen shot 2018-03-10 at 12 22 05](https://user-images.githubusercontent.com/9582/37242005-5df1066c-2462-11e8-9db2-aa2c709071b1.png)

![screen shot 2018-03-10 at 12 22 17](https://user-images.githubusercontent.com/9582/37242006-6167419e-2462-11e8-9373-7b2f0d968e80.png)

![screen shot 2018-03-10 at 12 22 34](https://user-images.githubusercontent.com/9582/37242008-68c91994-2462-11e8-8d33-edbefab48b65.png)

![screen shot 2018-03-10 at 12 23 36](https://user-images.githubusercontent.com/9582/37242009-6c4e700a-2462-11e8-8b17-7d4ea5515227.png)

The header and alerts were based on the PR by @inoas.

@rockwood, if you have suggestions, please let us know!

**Comment by webdeb** [comment]
The logo is also very nice, may I ask if it played a significant role for the final decision? ;-) 

![bildschirmfoto 2018-03-10 um 13 34 42](https://user-images.githubusercontent.com/14992140/37242334-18c4a930-2468-11e8-8638-3fdcdf114bef.png)

**Comment by inoas** [comment]
The font choices Milligram makes are terrible though. And not some of them - all of them: Kerning and font weights (too much/too little) as well as font sizes.

**Comment by josevalim** [comment]
@inoas is it something that we can easily change without overriding all definitions?

**Comment by inoas** [comment]
I got problems getting phoenixframework/master to run as demo locally (as a newbie):
But this could help:

Roboto:
<img width="1680" alt="roboto_1" src="https://user-images.githubusercontent.com/20972207/37244713-6df13bec-248d-11e8-9e58-578f5ce103c9.png">

<img width="1680" alt="roboto_2" src="https://user-images.githubusercontent.com/20972207/37244714-6e13a9fc-248d-11e8-97e6-c7ec58517040.png">


Without Roboto AND without Helvetica Neue AND without letter-spacing (bad practise globally/on body as it breaks/may break default font kerning)

<img width="1680" alt="helvetica_1" src="https://user-images.githubusercontent.com/20972207/37244716-6e5dccc6-248d-11e8-8c73-fba7e002b912.png">

<img width="1680" alt="helvetica_2" src="https://user-images.githubusercontent.com/20972207/37244715-6e40c78e-248d-11e8-8a27-2d10cb8b0182.png">

**Comment by josevalim** [comment]
Done in master. I do like it more too.

**Comment by glhrmv** [comment]
I feel like a classless CSS file is the best approach. It was mentioned that optimizing for removal is essential, so a single `phoenix.css` with Bootstrap made sense. But this ignores the consequences in the markup that including a class-based framework implies: it is not particularly fun having to delete the Bootstrap-specific markup, especially with all the ugly `div` nesting that it encourages by design.

I don't see how Milligram helps this issue, it seems like it will only make it more difficult for newer users by virtue of being much less known than Bootstrap, which is widely used.
